bblocks/import_tools/weo.py

```diff
@@ -0,0 +1,287 @@
+
+    def get_files(self):
+        """Get the root of files"""
+
+        if len(self.folder.namelist()) > 2:
+            raise ValueError("More than one file in zip file")
+
+        for file in self.folder.namelist():
+            if file.endswith(".xml"):
+                self.data_file = ET.parse(self.folder.open(file)).getroot()
+
+            if file.endswith(".xsd"):
+                self.schema_file = ET.parse(self.folder.open(file)).getroot()
+
+        if self.data_file is None or self.schema_file is None:
+            raise ValueError("No data or schema file found")
+
+    def extract_data(self):
+        """extract data from data file"""
+
+        rows = []
+        for series in self.data_file[1].findall("./"):
+            for obs in series.findall("./"):
+                rows.append({**series.attrib, **obs.attrib})
+
+        self.data = pd.DataFrame(rows)
+
+    def _convert_series_codes(self, series: pd.Series, lookup_value: str):
+        """ """
+
+        query = self.schema_file.findall(
+            "./{http://www.w3.org/2001/XMLSchema}simpleType[@name="
+            + f'"{lookup_value}"'
+            + "]/"
+        )[0].findall("./")
+
+        return series.map({elem.attrib["value"]: elem[0][0].text for elem in query})
+
+    def clean_data(self):
+        """Format the dataframe"""
+
+        for col, value in COLUMN_MAPPER.items():
+            self.data = self.data.rename(columns={col: f"{col}_CODE"})
+            self.data[col] = self._convert_series_codes(self.data[f"{col}_CODE"], value)
+
+        for col in ["REF_AREA_CODE", "LASTACTUALDATE", "TIME_PERIOD", "OBS_VALUE"]:
+            self.data[col] = pd.to_numeric(self.data[col], errors="coerce")
+
+    def parse_data(self) -> pd.DataFrame:
+        """Parse the data to a dataframe"""
+
+        self.get_files()
+        self.extract_data()
+        self.clean_data()
+
+        return self.data
+
+
+def gen_latest_version() -> tuple[int, int]:
+    """Generates the latest expected version based on the current date"""
+
+    current_year = datetime.now().year
+    current_month = datetime.now().month
+
+    # if month is less than 4 (April)
+    # return the version 2 (October) for the previous year
+    if current_month < 4:
+        current_year -= 1
+        return current_year, 2
+
+    # elif month is less than 10 (October)
+    # return current year and version 2 (April)
+    elif current_month < 10:
+        return current_year, 1
+
+    # else (if month is more than 10 (October)
+    # return current month and version 2 (October)
+    else:
+        return current_year, 2
+
+
+def roll_back_version(version: tuple[int, int]) -> tuple[int, int]:
+    """Roll back version to the previous version"""
+
+    if version[1] == 2:
+        return version[0], 1
+
+    elif version[1] == 1:
+        return version[0] - 1, 2
+
+    else:
+        raise ValueError(
+            f"Release must be either 1 or 2. Invalid release: {version[1]}"
+        )
```

Comment:
I'm not sure I get what this is supposed to do. (Maybe it will be clearer later)... but it would help to have a docstring that explains a bit of the purpose (as well as the expected behaviour like you have it now)

Reply:
the purpose is to roll back when trying to determine the latest release. Eg if it is October, but the data hasn't come out yet, this function facilitates rolling back the version to April